Data/Hashable.hs

```diff
@@ -231,13 +239,36 @@ instance Eq a => Eq (Hashed a) where
 instance Ord a => Ord (Hashed a) where
   Hashed a _ `compare` Hashed b _ = a `compare` b
 
-instance Hashable a => Hashable (Hashed a) where
+instance Show a => Show (Hashed a) where
+  showsPrec d (Hashed a _) = showParen (d > 10) $
+    showString "hashed" . showChar ' ' . showsPrec 11 a
+
+instance Hashable (Hashed a) where
   hashWithSalt = defaultHashWithSalt
   hash (Hashed _ h) = h
 
+-- This instance is a little unsettling. It is unusal for
+-- 'liftHashWithSalt' to ignore its first argument when a
+-- value is actually available for it to work on.
+instance Hashable1 Hashed where
+  liftHashWithSalt _ s (Hashed _ h) = defaultHashWithSalt s h
```

Comment:
It's a bit unsettling indeed. I wonder if `Hashed` to cache the salt it used last time and make a comparison to it and rehash if it's not the same.